File: dsa/python/strings_sliding_window.py
"""
Strings and Sliding Window
==========================

Common company-style questions:
1. Valid Anagram
2. Group Anagrams
3. Longest Substring Without Repeating Characters
4. Minimum Window Substring
5. Longest Palindromic Substring
6. Valid Palindrome
7. Longest Common Prefix
8. Find All Anagrams in a String
9. Permutation in String
10. Count Palindromic Substrings
11. Encode and Decode Strings

This topic tests:
- frequency maps
- pointer movement
- substring optimization

Quick input/output examples:
- `valid_anagram_count("listen", "silent") -> True`
- `group_anagrams(["eat", "tea", "ate"]) -> [["eat", "tea", "ate"]]`
- `longest_unique_substring_window("abcabcbb") -> 3`
- `minimum_window_substring("ADOBECODEBANC", "ABC") -> "BANC"`
- `longest_palindromic_substring("babad") -> "bab" or "aba"`
- `valid_palindrome("A man, a plan, a canal: Panama") -> True`
- `longest_common_prefix(["flower", "flow", "flight"]) -> "fl"`
- `find_all_anagrams("cbaebabacd", "abc") -> [0, 6]`
- `permutation_in_string("ab", "eidbaooo") -> True`
- `count_palindromic_substrings("aaa") -> 6`
- `encode_strings(["leet", "code"]) -> "4#leet4#code"`
- `decode_string("3[a2[c]]") -> "accaccacc"`
"""

from collections import Counter, defaultdict


def valid_anagram_sort(s: str, t: str) -> bool:
    """
    Approach:
    - Sort both strings
    - Compare them

    Complexity:
    - Time: O(n log n)
    - Space: O(n)

    Debugging steps:
    1. Print sorted versions of both strings
    2. Check lengths before sorting
    """
    return sorted(s) == sorted(t)


def valid_anagram_count(s: str, t: str) -> bool:
    """
    Better approach:
    - Count characters in both strings
    - Compare counters

    Complexity:
    - Time: O(n)
    - Space: O(n)
    """
    return Counter(s) == Counter(t)


def group_anagrams(words: list[str]) -> list[list[str]]:
    """
    Group words that are anagrams of each other.

    Approach:
    - Use sorted word as key
    - Append original word to that group

    Complexity:
    - Time: O(n * k log k)
    - Space: O(n * k)

    Debugging steps:
    1. Print each word and its normalized key
    2. Print the map after insertion
    """
    groups: dict[str, list[str]] = {}
    for word in words:
        key = "".join(sorted(word))
        groups.setdefault(key, []).append(word)
    return list(groups.values())


def longest_unique_substring_bruteforce(text: str) -> int:
    """
    Approach:
    - Check every substring
    - Keep the longest one with unique characters

    Complexity:
    - Time: O(n^3)
    - Space: O(n)

    Debugging steps:
    1. Print current substring
    2. Print set length versus substring length
    """
    best = 0
    for left in range(len(text)):
        for right in range(left, len(text)):
            current = text[left : right + 1]
            if len(set(current)) == len(current):
                best = max(best, len(current))
    return best


def longest_unique_substring_window(text: str) -> int:
    """
    Optimized sliding window:
    - Expand right pointer
    - Shrink left pointer until the window becomes unique

    Complexity:
    - Time: O(n)
    - Space: O(n)

    Debugging steps:
    1. Print left, right, current character
    2. Print window counts when shrinking
    3. Track when best answer changes
    """
    seen: dict[str, int] = {}
    left = 0
    best = 0

    for right, char in enumerate(text):
        if char in seen and seen[char] >= left:
            left = seen[char] + 1
        seen[char] = right
        best = max(best, right - left + 1)

    return best


def minimum_window_substring(source: str, target: str) -> str:
    """
    Problem:
    Find the smallest substring of source containing all characters of target.

    Core idea:
    - need = required character counts
    - have = counts inside current window
    - Expand right until valid
    - Shrink left while still valid

    Complexity:
    - Time: O(n)
    - Space: O(n)

    Debugging steps:
    1. Print need and window maps
    2. Track formed count against required count
    3. Print best window updates
    """
    if not source or not target:
        return ""

    need = Counter(target)
    window = defaultdict(int)
    formed = 0
    required = len(need)
    left = 0
    best_length = float("inf")
    best_range = (0, 0)

    for right, char in enumerate(source):
        window[char] += 1
        if char in need and window[char] == need[char]:
            formed += 1

        while formed == required:
            if right - left + 1 < best_length:
                best_length = right - left + 1
                best_range = (left, right)

            left_char = source[left]
            window[left_char] -= 1
            if left_char in need and window[left_char] < need[left_char]:
                formed -= 1
            left += 1

    if best_length == float("inf"):
        return ""
    start, end = best_range
    return source[start : end + 1]


def longest_palindromic_substring(text: str) -> str:
    """
    Expand around every possible center.

    Complexity:
    - Time: O(n^2)
    - Space: O(1)

    Debugging steps:
    1. Print left and right center
    2. Print substring while expanding
    """
    if not text:
        return ""

    best = ""

    def expand(left: int, right: int) -> str:
        while left >= 0 and right < len(text) and text[left] == text[right]:
            left -= 1
            right += 1
        return text[left + 1 : right]

    for index in range(len(text)):
        odd = expand(index, index)
        even = expand(index, index + 1)
        current = odd if len(odd) > len(even) else even
        if len(current) > len(best):
            best = current

    return best


def valid_palindrome(text: str) -> bool:
    """
    Ignore non-alphanumeric characters and compare from both ends.
    """
    left = 0
    right = len(text) - 1

    while left < right:
        while left < right and not text[left].isalnum():
            left += 1
        while left < right and not text[right].isalnum():
            right -= 1

        if text[left].lower() != text[right].lower():
            return False

        left += 1
        right -= 1

    return True


def longest_common_prefix(words: list[str]) -> str:
    """
    Compare characters position by position.
    """
    if not words:
        return ""

    prefix = words[0]
    for word in words[1:]:
        while not word.startswith(prefix):
            prefix = prefix[:-1]
            if not prefix:
                return ""
    return prefix


def find_all_anagrams(source: str, target: str) -> list[int]:
    """
    Return all start indices where an anagram of target appears.
    """
    if len(target) > len(source):
        return []

    target_count = Counter(target)
    window = Counter(source[: len(target)])
    result: list[int] = []

    if window == target_count:
        result.append(0)

    for right in range(len(target), len(source)):
        window[source[right]] += 1
        left_char = source[right - len(target)]
        window[left_char] -= 1
        if window[left_char] == 0:
            del window[left_char]
        if window == target_count:
            result.append(right - len(target) + 1)

    return result


def permutation_in_string(pattern: str, text: str) -> bool:
    """
    Check whether any permutation of pattern exists in text.
    """
    return len(find_all_anagrams(text, pattern)) > 0


def count_palindromic_substrings(text: str) -> int:
    """
    Count all palindromic substrings by expanding around centers.
    """
    count = 0

    def expand(left: int, right: int) -> int:
        local_count = 0
        while left >= 0 and right < len(text) and text[left] == text[right]:
            local_count += 1
            left -= 1
            right += 1
        return local_count

    for index in range(len(text)):
        count += expand(index, index)
        count += expand(index, index + 1)

    return count


def encode_strings(strings: list[str]) -> str:
    """
    Encode a list of strings using length prefixes.
    """
    return "".join(f"{len(item)}#{item}" for item in strings)


def decode_strings(encoded: str) -> list[str]:
    """
    Decode a length-prefixed encoded string.
    """
    result: list[str] = []
    index = 0

    while index < len(encoded):
        separator = index
        while encoded[separator] != "#":
            separator += 1
        length = int(encoded[index:separator])
        start = separator + 1
        end = start + length
        result.append(encoded[start:end])
        index = end

    return result


def longest_repeating_character_replacement(text: str, k: int) -> int:
    """
    Find the longest window that can become one repeating character
    after at most k replacements.
    """
    counts: dict[str, int] = defaultdict(int)
    left = 0
    best = 0
    max_frequency = 0

    for right, char in enumerate(text):
        counts[char] += 1
        max_frequency = max(max_frequency, counts[char])

        while (right - left + 1) - max_frequency > k:
            counts[text[left]] -= 1
            left += 1

        best = max(best, right - left + 1)

    return best


def zigzag_conversion(text: str, num_rows: int) -> str:
    """
    Arrange text in zigzag order and read row by row.
    """
    if num_rows == 1 or num_rows >= len(text):
        return text

    rows = [""] * num_rows
    current_row = 0
    direction = 1

    for char in text:
        rows[current_row] += char
        if current_row == 0:
            direction = 1
        elif current_row == num_rows - 1:
            direction = -1
        current_row += direction

    return "".join(rows)


def string_to_integer_atoi(text: str) -> int:
    """
    Parse string into integer while handling spaces and sign.
    """
    text = text.lstrip()
    if not text:
        return 0

    sign = 1
    index = 0
    if text[index] in "+-":
        sign = -1 if text[index] == "-" else 1
        index += 1

    result = 0
    while index < len(text) and text[index].isdigit():
        result = result * 10 + int(text[index])
        index += 1

    result *= sign
    lower_bound = -(2**31)
    upper_bound = 2**31 - 1
    return max(lower_bound, min(upper_bound, result))


def implement_strstr(haystack: str, needle: str) -> int:
    """
    Return the first index of needle in haystack.
    """
    if needle == "":
        return 0

    for start in range(len(haystack) - len(needle) + 1):
        if haystack[start : start + len(needle)] == needle:
            return start
    return -1


def decode_string(text: str) -> str:
    """
    Decode nested repeat expressions like 3[a2[c]].
    """
    stack: list[tuple[str, int]] = []
    current_string = ""
    current_number = 0

    for char in text:
        if char.isdigit():
            current_number = current_number * 10 + int(char)
        elif char == "[":
            stack.append((current_string, current_number))
            current_string = ""
            current_number = 0
        elif char == "]":
            previous_string, repeat_count = stack.pop()
            current_string = previous_string + current_string * repeat_count
        else:
            current_string += char

    return current_string


def substring_with_concatenation_of_all_words(text: str, words: list[str]) -> list[int]:
    """
    Find start indices where all words appear exactly once without gaps.
    """
    if not text or not words:
        return []

    word_length = len(words[0])
    total_words = len(words)
    total_length = word_length * total_words
    target = Counter(words)
    result: list[int] = []

    for offset in range(word_length):
        left = offset
        seen: dict[str, int] = defaultdict(int)
        used_words = 0

        for right in range(offset, len(text) - word_length + 1, word_length):
            word = text[right : right + word_length]
            if word in target:
                seen[word] += 1
                used_words += 1

                while seen[word] > target[word]:
                    left_word = text[left : left + word_length]
                    seen[left_word] -= 1
                    used_words -= 1
                    left += word_length

                if used_words == total_words:
                    result.append(left)
                    left_word = text[left : left + word_length]
                    seen[left_word] -= 1
                    used_words -= 1
                    left += word_length
            else:
                seen.clear()
                used_words = 0
                left = right + word_length

    valid_limit = len(text) - total_length
    return [index for index in result if index <= valid_limit]


if __name__ == "__main__":
    print("Valid Anagram:", valid_anagram_count("listen", "silent"))
    print("Group Anagrams:", group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
    print("Longest Unique Substring:", longest_unique_substring_window("abcabcbb"))
    print("Minimum Window Substring:", minimum_window_substring("ADOBECODEBANC", "ABC"))
    print("Longest Palindromic Substring:", longest_palindromic_substring("babad"))
    print("Valid Palindrome:", valid_palindrome("A man, a plan, a canal: Panama"))
    print("Longest Common Prefix:", longest_common_prefix(["flower", "flow", "flight"]))
    print("Find All Anagrams:", find_all_anagrams("cbaebabacd", "abc"))
    print("Permutation In String:", permutation_in_string("ab", "eidbaooo"))
    print("Count Palindromic Substrings:", count_palindromic_substrings("aaa"))
    encoded = encode_strings(["leet", "code", "python"])
    print("Encode Strings:", encoded)
    print("Decode Strings:", decode_strings(encoded))
    print("Longest Repeating Character Replacement:", longest_repeating_character_replacement("AABABBA", 1))
    print("Zigzag Conversion:", zigzag_conversion("PAYPALISHIRING", 3))
    print("String to Integer:", string_to_integer_atoi("   -42"))
    print("Implement strStr:", implement_strstr("sadbutsad", "sad"))
    print("Decode String:", decode_string("3[a2[c]]"))
    print(
        "Substring With Concatenation:",
        substring_with_concatenation_of_all_words("barfoothefoobarman", ["foo", "bar"]),
    )

File: dsa/javascript/strings_sliding_window.js
/*
Strings and Sliding Window
==========================

Common company-style questions:
1. Valid Anagram
2. Group Anagrams
3. Longest Substring Without Repeating Characters
4. Minimum Window Substring
5. Longest Palindromic Substring
6. Valid Palindrome
7. Longest Common Prefix
8. Find All Anagrams in a String
9. Permutation in String
10. Count Palindromic Substrings
11. Encode and Decode Strings

Quick input/output examples:
- validAnagramCount("listen", "silent") -> true
- groupAnagrams(["eat", "tea", "ate"]) -> [["eat", "tea", "ate"]]
- longestUniqueSubstringWindow("abcabcbb") -> 3
- minimumWindowSubstring("ADOBECODEBANC", "ABC") -> "BANC"
- longestPalindromicSubstring("babad") -> "bab" or "aba"
- validPalindrome("A man, a plan, a canal: Panama") -> true
- longestCommonPrefix(["flower", "flow", "flight"]) -> "fl"
- findAllAnagrams("cbaebabacd", "abc") -> [0, 6]
- permutationInString("ab", "eidbaooo") -> true
- countPalindromicSubstrings("aaa") -> 6
- encodeStrings(["leet", "code"]) -> "4#leet4#code"
- decodeString("3[a2[c]]") -> "accaccacc"
*/

function validAnagramSort(s, t) {
  return [...s].sort().join("") === [...t].sort().join("");
}

function validAnagramCount(s, t) {
  if (s.length !== t.length) {
    return false;
  }

  const count = new Map();
  for (const char of s) {
    count.set(char, (count.get(char) || 0) + 1);
  }

  for (const char of t) {
    if (!count.has(char)) {
      return false;
    }
    count.set(char, count.get(char) - 1);
    if (count.get(char) === 0) {
      count.delete(char);
    }
  }

  return count.size === 0;
}

function groupAnagrams(words) {
  const groups = new Map();

  for (const word of words) {
    const key = [...word].sort().join("");
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key).push(word);
  }

  return [...groups.values()];
}

function longestUniqueSubstringWindow(text) {
  /*
  Debugging steps:
  1. Print left, right, current char
  2. Print map after updates
  */
  const seen = new Map();
  let left = 0;
  let best = 0;

  for (let right = 0; right < text.length; right += 1) {
    const char = text[right];
    if (seen.has(char) && seen.get(char) >= left) {
      left = seen.get(char) + 1;
    }
    seen.set(char, right);
    best = Math.max(best, right - left + 1);
  }

  return best;
}

function minimumWindowSubstring(source, target) {
  if (!source || !target) {
    return "";
  }

  const need = new Map();
  for (const char of target) {
    need.set(char, (need.get(char) || 0) + 1);
  }

  const window = new Map();
  let formed = 0;
  const required = need.size;
  let left = 0;
  let bestLength = Infinity;
  let bestStart = 0;

  for (let right = 0; right < source.length; right += 1) {
    const char = source[right];
    window.set(char, (window.get(char) || 0) + 1);

    if (need.has(char) && window.get(char) === need.get(char)) {
      formed += 1;
    }

    while (formed === required) {
      if (right - left + 1 < bestLength) {
        bestLength = right - left + 1;
        bestStart = left;
      }

      const leftChar = source[left];
      window.set(leftChar, window.get(leftChar) - 1);
      if (need.has(leftChar) && window.get(leftChar) < need.get(leftChar)) {
        formed -= 1;
      }
      left += 1;
    }
  }

  return bestLength === Infinity ? "" : source.slice(bestStart, bestStart + bestLength);
}

function longestPalindromicSubstring(text) {
  if (!text) {
    return "";
  }

  let best = "";

  function expand(left, right) {
    while (left >= 0 && right < text.length && text[left] === text[right]) {
      left -= 1;
      right += 1;
    }
    return text.slice(left + 1, right);
  }

  for (let index = 0; index < text.length; index += 1) {
    const odd = expand(index, index);
    const even = expand(index, index + 1);
    const current = odd.length > even.length ? odd : even;
    if (current.length > best.length) {
      best = current;
    }
  }

  return best;
}

function validPalindrome(text) {
  let left = 0;
  let right = text.length - 1;

  while (left < right) {
    while (left < right && !/[a-z0-9]/i.test(text[left])) {
      left += 1;
    }
    while (left < right && !/[a-z0-9]/i.test(text[right])) {
      right -= 1;
    }

    if (text[left].toLowerCase() !== text[right].toLowerCase()) {
      return false;
    }

    left += 1;
    right -= 1;
  }

  return true;
}

function longestCommonPrefix(words) {
  if (!words.length) {
    return "";
  }

  let prefix = words[0];
  for (let index = 1; index < words.length; index += 1) {
    while (!words[index].startsWith(prefix)) {
      prefix = prefix.slice(0, -1);
      if (!prefix) {
        return "";
      }
    }
  }

  return prefix;
}

function findAllAnagrams(source, target) {
  if (target.length > source.length) {
    return [];
  }

  const need = new Map();
  const window = new Map();
  const result = [];

  for (const char of target) {
    need.set(char, (need.get(char) || 0) + 1);
  }

  function sameCounts() {
    if (need.size !== window.size) {
      return false;
    }
    for (const [key, value] of need.entries()) {
      if (window.get(key) !== value) {
        return false;
      }
    }
    return true;
  }

  for (let index = 0; index < target.length; index += 1) {
    const char = source[index];
    window.set(char, (window.get(char) || 0) + 1);
  }

  if (sameCounts()) {
    result.push(0);
  }

  for (let right = target.length; right < source.length; right += 1) {
    const addChar = source[right];
    window.set(addChar, (window.get(addChar) || 0) + 1);

    const removeChar = source[right - target.length];
    window.set(removeChar, window.get(removeChar) - 1);
    if (window.get(removeChar) === 0) {
      window.delete(removeChar);
    }

    if (sameCounts()) {
      result.push(right - target.length + 1);
    }
  }

  return result;
}

function permutationInString(pattern, text) {
  return findAllAnagrams(text, pattern).length > 0;
}

function countPalindromicSubstrings(text) {
  let count = 0;

  function expand(left, right) {
    let localCount = 0;
    while (left >= 0 && right < text.length && text[left] === text[right]) {
      localCount += 1;
      left -= 1;
      right += 1;
    }
    return localCount;
  }

  for (let index = 0; index < text.length; index += 1) {
    count += expand(index, index);
    count += expand(index, index + 1);
  }

  return count;
}

function encodeStrings(strings) {
  return strings.map((item) => `${item.length}#${item}`).join("");
}

function decodeStrings(encoded) {
  const result = [];
  let index = 0;

  while (index < encoded.length) {
    let separator = index;
    while (encoded[separator] !== "#") {
      separator += 1;
    }
    const length = Number(encoded.slice(index, separator));
    const start = separator + 1;
    const end = start + length;
    result.push(encoded.slice(start, end));
    index = end;
  }

  return result;
}

function longestRepeatingCharacterReplacement(text, k) {
  const counts = new Map();
  let left = 0;
  let best = 0;
  let maxFrequency = 0;

  for (let right = 0; right < text.length; right += 1) {
    const char = text[right];
    counts.set(char, (counts.get(char) || 0) + 1);
    maxFrequency = Math.max(maxFrequency, counts.get(char));

    while (right - left + 1 - maxFrequency > k) {
      const leftChar = text[left];
      counts.set(leftChar, counts.get(leftChar) - 1);
      left += 1;
    }

    best = Math.max(best, right - left + 1);
  }

  return best;
}

function zigzagConversion(text, numRows) {
  if (numRows === 1 || numRows >= text.length) {
    return text;
  }

  const rows = new Array(numRows).fill("");
  let currentRow = 0;
  let direction = 1;

  for (const char of text) {
    rows[currentRow] += char;
    if (currentRow === 0) {
      direction = 1;
    } else if (currentRow === numRows - 1) {
      direction = -1;
    }
    currentRow += direction;
  }

  return rows.join("");
}

function stringToIntegerAtoi(text) {
  text = text.trimStart();
  if (!text) {
    return 0;
  }

  let sign = 1;
  let index = 0;
  if (text[index] === "+" || text[index] === "-") {
    sign = text[index] === "-" ? -1 : 1;
    index += 1;
  }

  let result = 0;
  while (index < text.length && /\d/.test(text[index])) {
    result = result * 10 + Number(text[index]);
    index += 1;
  }

  result *= sign;
  const lowerBound = -(2 ** 31);
  const upperBound = 2 ** 31 - 1;
  return Math.max(lowerBound, Math.min(upperBound, result));
}

function implementStrStr(haystack, needle) {
  if (needle === "") {
    return 0;
  }

  for (let start = 0; start <= haystack.length - needle.length; start += 1) {
    if (haystack.slice(start, start + needle.length) === needle) {
      return start;
    }
  }
  return -1;
}

function decodeString(text) {
  const stack = [];
  let currentString = "";
  let currentNumber = 0;

  for (const char of text) {
    if (/\d/.test(char)) {
      currentNumber = currentNumber * 10 + Number(char);
    } else if (char === "[") {
      stack.push([currentString, currentNumber]);
      currentString = "";
      currentNumber = 0;
    } else if (char === "]") {
      const [previousString, repeatCount] = stack.pop();
      currentString = previousString + currentString.repeat(repeatCount);
    } else {
      currentString += char;
    }
  }

  return currentString;
}

function substringWithConcatenationOfAllWords(text, words) {
  if (!text || !words.length) {
    return [];
  }

  const wordLength = words[0].length;
  const totalWords = words.length;
  const totalLength = wordLength * totalWords;
  const target = new Map();
  const result = [];

  for (const word of words) {
    target.set(word, (target.get(word) || 0) + 1);
  }

  for (let offset = 0; offset < wordLength; offset += 1) {
    let left = offset;
    const seen = new Map();
    let usedWords = 0;

    for (let right = offset; right <= text.length - wordLength; right += wordLength) {
      const word = text.slice(right, right + wordLength);

      if (target.has(word)) {
        seen.set(word, (seen.get(word) || 0) + 1);
        usedWords += 1;

        while (seen.get(word) > target.get(word)) {
          const leftWord = text.slice(left, left + wordLength);
          seen.set(leftWord, seen.get(leftWord) - 1);
          usedWords -= 1;
          left += wordLength;
        }

        if (usedWords === totalWords) {
          result.push(left);
          const leftWord = text.slice(left, left + wordLength);
          seen.set(leftWord, seen.get(leftWord) - 1);
          usedWords -= 1;
          left += wordLength;
        }
      } else {
        seen.clear();
        usedWords = 0;
        left = right + wordLength;
      }
    }
  }

  return result.filter((index) => index <= text.length - totalLength);
}

console.log("Valid Anagram:", validAnagramCount("listen", "silent"));
console.log("Group Anagrams:", groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]));
console.log("Longest Unique Substring:", longestUniqueSubstringWindow("abcabcbb"));
console.log("Minimum Window:", minimumWindowSubstring("ADOBECODEBANC", "ABC"));
console.log("Longest Palindromic Substring:", longestPalindromicSubstring("babad"));
console.log("Valid Palindrome:", validPalindrome("A man, a plan, a canal: Panama"));
console.log("Longest Common Prefix:", longestCommonPrefix(["flower", "flow", "flight"]));
console.log("Find All Anagrams:", findAllAnagrams("cbaebabacd", "abc"));
console.log("Permutation In String:", permutationInString("ab", "eidbaooo"));
console.log("Count Palindromic Substrings:", countPalindromicSubstrings("aaa"));
const encoded = encodeStrings(["leet", "code", "python"]);
console.log("Encode Strings:", encoded);
console.log("Decode Strings:", decodeStrings(encoded));
console.log("Longest Repeating Character Replacement:", longestRepeatingCharacterReplacement("AABABBA", 1));
console.log("Zigzag Conversion:", zigzagConversion("PAYPALISHIRING", 3));
console.log("String to Integer:", stringToIntegerAtoi("   -42"));
console.log("Implement strStr:", implementStrStr("sadbutsad", "sad"));
console.log("Decode String:", decodeString("3[a2[c]]"));
console.log(
  "Substring With Concatenation:",
  substringWithConcatenationOfAllWords("barfoothefoobarman", ["foo", "bar"])
);
